python_modules/dagster/dagster/_daemon/monitoring/monitoring_daemon.py

```diff
@@ -102,34 +110,83 @@ def execute_monitoring_iteration(
 
     # TODO: consider limiting number of runs to fetch
-    runs = list(instance.get_runs(filters=RunsFilter(statuses=IN_PROGRESS_RUN_STATUSES)))
+    run_records = list(
+        instance.get_run_records(filters=RunsFilter(statuses=IN_PROGRESS_RUN_STATUSES))
+    )
 
-    if not runs:
+    if not run_records:
         return
 
-    logger.info(f"Collected {len(runs)} runs for monitoring")
+    logger.info(f"Collected {len(run_records)} runs for monitoring")
     workspace = workspace_process_context.create_request_context()
-    for run in runs:
+    for run_record in run_records:
         try:
-            logger.info(f"Checking run {run.run_id}")
+            logger.info(f"Checking run {run_record.dagster_run.run_id}")
 
-            if run.status == DagsterRunStatus.STARTING:
-                monitor_starting_run(instance, run, logger)
-            elif run.status == DagsterRunStatus.STARTED:
-                monitor_started_run(instance, workspace, run, logger)
-            elif run.status == DagsterRunStatus.CANCELING:
+            if run_record.dagster_run.status == DagsterRunStatus.STARTING:
+                monitor_starting_run(instance, run_record, logger)
+            elif run_record.dagster_run.status == DagsterRunStatus.STARTED:
+                monitor_started_run(instance, workspace, run_record, logger)
+            elif run_record.dagster_run.status == DagsterRunStatus.CANCELING:
                 # TODO: implement canceling timeouts
                 pass
             else:
-                check.invariant(False, f"Unexpected run status: {run.status}")
+                check.invariant(False, f"Unexpected run status: {run_record.dagster_run.status}")
         except Exception:
             error_info = serializable_error_info_from_exc_info(sys.exc_info())
-            logger.error(f"Hit error while monitoring run {run.run_id}: {str(error_info)}")
+            logger.error(
+                f"Hit error while monitoring run {run_record.dagster_run.run_id}: {str(error_info)}"
+            )
             yield error_info
         else:
             yield
+
+
+def check_run_timeout(
+    instance: DagsterInstance, run_record: RunRecord, logger: logging.Logger
+) -> None:
+    max_time_str = run_record.dagster_run.tags.get(
+        MAX_RUNTIME_TAG,
+    )
+    if not max_time_str:
+        return
+
+    max_time = float(max_time_str)
+
+    if (
+        run_record.start_time is not None
+        and pendulum.now("UTC").timestamp() - run_record.start_time > max_time
+    ):
+        logger.info(
+            f"Run {run_record.dagster_run.run_id} has exceeded maximum runtime of"
+            f" {max_time} seconds: terminating run."
+        )
+
+        instance.report_run_canceling(
+            run_record.dagster_run,
+            message=f"Canceling due to exceeding maximum runtime of {max_time} seconds.",
+        )
+        try:
+            instance.run_launcher.terminate(run_id=run_record.dagster_run.run_id)
+        except:
```

Comment:
i think we should make this work more like the FORCE_TERMINATE_IMMEDIATELY case in graphql and always move it into a CANCELED state no matter what (while still trying to terminate it cleanly if possible) - we shouldn't ever leave runs in CANCELING when this happens